src/stream/src/executor/utils.rs

```diff
@@ -22,3 +22,21 @@ impl Execute for DummyExecutor {
         futures::stream::pending().boxed()
     }
 }
+
+pub fn compute_rate_limit_chunk_permits(chunk: &StreamChunk, limit: usize) -> usize {
+    let chunk_size = chunk.capacity();
+    let ends_with_update = if chunk_size >= 2 {
+        // Note we have to check if the 2nd last is `U-` to be consistenct with `StreamChunkBuilder`.
+        // If something inconsistent happens in the stream, we may not have `U+` after this `U-`.
+        chunk.ops()[chunk_size - 2].is_update_delete()
+    } else {
+        false
+    };
+    if chunk_size == limit + 1 && ends_with_update {
```

Comment:
What if the limit is `1` and the chunk is [-U, +U]

Reply:
The chunk of size 2 can successfully obtain 1 permit from the rate limiter.
Actually this is the code snippet from https://github.com/risingwavelabs/risingwave/pull/16407. It's been moved to utils.rs for reusability.